### PR TITLE
[ConvectionDiffusionApplication] Modernize application. Adding solver and analysis stage (up-to-date to the master)

### PR DESCRIPTION
**Pull request by loumalouomega**
Equivalent to #2224, but for the master branch. Adapted to the last Stage and Base solver

### CODE REVIEW

**Review by RiccardoRossi — DISMISSED** [review]
comments were made in the equivalent PR for release.

@loumalouomega  the comments i made in the other branch should be addressed prior to merging this one

**Comment by loumalouomega** [comment]
Those changes are addressed and more, check the code

File: applications/convection_diffusion_application/custom_elements/eulerian_diff.h [hunk 1]
```diff
@@ -89,7 +55,7 @@ namespace Kratos
 ///@name Kratos Classes
 ///@{
 
-///formulation described in https://docs.google.com/document/d/13a_zGLj6xORDuLgoOG5LwHI6BwShvfO166opZ815zLY/edit?usp=sharing
+///formulation described in
```
Comment (RiccardoRossi) [inline applications/convection_diffusion_application/custom_elements/eulerian_diff.h]: why did u remove the link?

File: applications/convection_diffusion_application/license.txt [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+Copyright (c) 2018, Riccardo Rossi, Jordi Cotela Dalmau, CIMNE (International Center for Numerical Methods in Engineering)
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:
+
+	-	Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.
+	-	Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer 
+		in the documentation and/or other materials provided with the distribution.
+	-	All advertising materials mentioning features or use of this software must display the following acknowledgement: 
+			This product includes Kratos Structural Mechanics technology.
```
Comment (RiccardoRossi) [inline applications/convection_diffusion_application/license.txt]: Convection Diffusion Technology ;-)

File: applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py [hunk 3]
```diff
@@ -0,0 +1,757 @@
...
+            if not custom_settings["convection_diffusion_variables"].Has("diffusion_variable"):
+                self.print_warning_on_rank_zero("::[ConvectionDiffusionBaseSolver]:: ", "W-A-R-N-I-N-G: DIFUSSION VARIABLE NOT DEFINED, TAKING DEFAULT", default_settings["convection_diffusion_variables"]["diffusion_variable"].GetString())
+            if not custom_settings["convection_diffusion_variables"].Has("unknown_variable"):
+                self.print_warning_on_rank_zero("::[ConvectionDiffusionBaseSolver]:: ", "W-A-R-N-I-N-G: UNKNOWN VARIABLE NOT DEFINED, TAKING DEFAULT", default_settings["convection_diffusion_variables"]["unknown_variable"].GetString())
+            if not custom_settings["convection_diffusion_variables"].Has("volume_source_variable"):
+                self.print_warning_on_rank_zero("::[ConvectionDiffusionBaseSolver]:: ", "W-A-R-N-I-N-G: VOLUME SOURCE VARIABLE NOT DEFINED, TAKING DEFAULT", default_settings["convection_diffusion_variables"]["volume_source_variable"].GetString())
```
Comment (RiccardoRossi) [inline applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py:122]: are u sure that a default is defined? for example what would it be in this case?

File: applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py [hunk 4]
```diff
@@ -0,0 +1,757 @@
...
+        if not hasattr(self, '_linear_solver'):
+            self._linear_solver = self._create_linear_solver()
+        return self._linear_solver
+
+    def get_builder_and_solver(self):
+        if not hasattr(self, '_builder_and_solver'):
+            self._builder_and_solver = self._create_builder_and_solver()
+        return self._builder_and_solver
+
+    def get_mechanical_solution_strategy(self):
```
Comment (RiccardoRossi) [inline applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py]: why a mechanical solution strategy in a convection_diffusion one?

File: applications/convection_diffusion_application/python_scripts/convection_diffusion_stationary_solver.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+        super(ConvectionDiffusionStationarySolver, self).__init__(main_model_part, custom_settings)
+        self.print_on_rank_zero("::[ConvectionDiffusionStationarySolver]:: ", "Construction finished")
+
+    #### Private functions ####
+
+    def _create_solution_scheme(self):
+        #Variable defining the temporal scheme (0: Forward Euler, 1: Backward Euler, 0.5: Crank-Nicolson)
+        self.GetComputingModelPart().ProcessInfo[ConvectionDiffusionApplication.THETA] = 1.0
+        self.GetComputingModelPart().ProcessInfo[KratosMultiphysics.DYNAMIC_TAU] = 0.0
+        mechanical_scheme = KratosMultiphysics.ResidualBasedIncrementalUpdateStaticScheme()
```
Comment (RiccardoRossi) [inline applications/convection_diffusion_application/python_scripts/convection_diffusion_stationary_solver.py]: this should NOT be called mechanical_Scheme. call it scheme, convection_scheme, whatever...but not mechanical_*

File: applications/convection_diffusion_application/python_scripts/convection_diffusion_transient_solver.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+        super(ConvectionDiffusionTransientSolver, self).__init__(main_model_part, custom_settings)
+        self.print_on_rank_zero("::[ConvectionDiffusionTransientSolver]:: ", "Construction finished")
+
+    #### Private functions ####
+
+    def _create_solution_scheme(self):
+        #Variable defining the temporal scheme (0: Forward Euler, 1: Backward Euler, 0.5: Crank-Nicolson)
+        self.GetComputingModelPart().ProcessInfo[ConvectionDiffusionApplication.THETA] = self.transient_settings["transient_parameters"]["theta"].GetDouble()
+        self.GetComputingModelPart().ProcessInfo[KratosMultiphysics.DYNAMIC_TAU] = self.transient_settings["transient_parameters"]["dynamic_tau"].GetDouble()
+        mechanical_scheme = KratosMultiphysics.ResidualBasedIncrementalUpdateStaticScheme()
```
Comment (RiccardoRossi) [inline applications/convection_diffusion_application/python_scripts/convection_diffusion_transient_solver.py]: same

Comment (RiccardoRossi) [reply applications/convection_diffusion_application/python_scripts/convection_diffusion_transient_solver.py]: btw, DO NOT touch the Structural Mechanics, it is fine there...

File: applications/convection_diffusion_application/python_scripts/proposal_convection_diffusion_solver.py [hunk 7]
```diff
@@ -0,0 +1,271 @@
+from __future__ import print_function, absolute_import, division
```
Comment (RiccardoRossi) [inline applications/convection_diffusion_application/python_scripts/proposal_convection_diffusion_solver.py]: i already asked you what is this "proposal_" about...

**Comment by philbucher** [comment]
@loumalouomega be aware that the stages changes in #2135 !
Most likely it will be much simpler

File: applications/convection_diffusion_application/custom_elements/eulerian_diff.h (hunk 1 above)
Comment (loumalouomega) [reply applications/convection_diffusion_application/custom_elements/eulerian_diff.h]: Ups

File: applications/convection_diffusion_application/license.txt (hunk 2 above)
Comment (loumalouomega) [reply applications/convection_diffusion_application/license.txt]: Any better suggestion?

File: applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py (hunk 3 above)
Comment (loumalouomega) [reply applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py:122]: The ones in the default parameters

File: applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py (hunk 4 above)
Comment (loumalouomega) [reply applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py]: CP from structural, sorry

File: applications/convection_diffusion_application/license.txt (hunk 2 above)
Comment (jcotela) [reply applications/convection_diffusion_application/license.txt]: Do we need a separate license for the convection diffusion application? I'm partial to releasing as much as possible with the core Kratos license.

File: applications/convection_diffusion_application/python_scripts/proposal_convection_diffusion_solver.py (hunk 7 above)
Comment (loumalouomega) [reply applications/convection_diffusion_application/python_scripts/proposal_convection_diffusion_solver.py]: Is the one added by @jcotela, I restored the old one, because without it the old tests don't pass

File: applications/convection_diffusion_application/python_scripts/convection_diffusion_transient_solver.py [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+
+    See convection_diffusion_base_solver.py for more information.
+    """
+    def __init__(self, main_model_part, custom_settings):
+        # Set defaults and validate custom settings.
+        self.transient_settings = KratosMultiphysics.Parameters("""
+        {
+            "transient_parameters" : {
+                "dynamic_tau": 1.0,
+                "theta"    : 1.0
```
Comment (jcotela) [inline applications/convection_diffusion_application/python_scripts/convection_diffusion_transient_solver.py]: Why is Backward Euler the default? Crank-Nicolson (0.5) is second order...

File: applications/convection_diffusion_application/python_scripts/proposal_convection_diffusion_solver.py (hunk 7 above)
Comment (jcotela) [reply applications/convection_diffusion_application/python_scripts/proposal_convection_diffusion_solver.py]: Why do we even need it? You already added three completely new solvers. (BTW, some tests in the fluid dynamics are broken, since I changed them in #2214 to use my draft, and these changes have somehow ended up in the release-6.0 branch)

File: applications/convection_diffusion_application/python_scripts/convection_diffusion_transient_solver.py (hunk 8 above)
Comment (loumalouomega) [reply applications/convection_diffusion_application/python_scripts/convection_diffusion_transient_solver.py]: Well, it gave me less problems

Comment (loumalouomega) [reply applications/convection_diffusion_application/python_scripts/convection_diffusion_transient_solver.py]: I am not an expert

Comment (loumalouomega) [reply applications/convection_diffusion_application/python_scripts/convection_diffusion_transient_solver.py]: Done

**Comment by loumalouomega** [comment]
@RiccardoRossi and @jcotela done

**Comment by loumalouomega** [comment]
Ping

**Comment by philbucher** [comment]
I still recommend to wait for #2135 ...

**Comment by loumalouomega** [comment]
Ok, I thought we already merged that

**Comment by philbucher** [comment]
I am waiting for Jordi for making the Solvers in Fluid deriving from the base-solver
Without this #2135 would break the fluid...

**Comment by loumalouomega** [comment]
Ok

**Comment by loumalouomega** [comment]
Ping

**Comment by philbucher** [comment]
Still waiting for #2135 ...

**Comment by loumalouomega** [comment]
@philbucher Done, updated to the last standard

**Comment by loumalouomega** [comment]
Done @RiccardoRossi

**Comment by ipouplana** [comment]
Please @RiccardoRossi, can we merge this?, I need it for my work with @loumalouomega 

Regards

File: applications/convection_diffusion_application/license.txt (hunk 2 above)
Comment (RiccardoRossi) [reply applications/convection_diffusion_application/license.txt]: as Jordi is suggesting, let's use the general one for this (that is, let's just remove this file)

File: applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py [hunk 9]
```diff
@@ -0,0 +1,757 @@
...
+            "block_builder": true,
+            "clear_storage": false,
+            "move_mesh_flag": false,
+            "convergence_criterion": "residual_criterion",
+            "solution_relative_tolerance": 1.0e-4,
+            "solution_absolute_tolerance": 1.0e-9,
+            "residual_relative_tolerance": 1.0e-4,
+            "residual_absolute_tolerance": 1.0e-9,
+            "max_iteration": 10,
+            "linear_solver_settings":{
```
Comment (RiccardoRossi) [inline applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py:92]: i would use the amgcl solver as deafult solver. It will be much more robust than this choice

**Comment by RiccardoRossi** [comment]
Hi. this looks fine to me, apart for removing the specific license in favor of leaving the generic kratos one as @jcotela is suggesting.

I will remove my blocker, however:
1 - i would like approval to be given by @jcotela since i know that he is using this
2 - after merging this we need to verify if the coupling example in the tutorial still works and eventualyl updatate it.

also @philbucher does this look fine to you?

**Review by RiccardoRossi — COMMENTED** [review]
minor comments

**Comment by loumalouomega** [comment]
@RiccardoRossi done, changes suggested done

**Comment by philbucher** [comment]
I will take a close look tmr

File: applications/convection_diffusion_application/custom_conditions/flux_condition.h [hunk 10]
```diff
@@ -1,11 +1,10 @@
...
-//                   Multi-Physics
+// KRATOS ___ ___  _  ___   __   ___ ___ ___ ___ 
+//       / __/ _ \| \| \ \ / /__|   \_ _| __| __|
+//      | (_| (_) | .` |\ V /___| |) | || _|| _| 
+//       \___\___/|_|\_| \_/    |___/___|_| |_|  APPLICATION
 //
-//  License:         BSD License
-//                   Kratos default license: kratos/license.txt
+//  License: BSD License
+//   license: convection_diffusion_application/license.txt
```
Comment (philbucher) [inline applications/convection_diffusion_application/custom_conditions/flux_condition.h]: didn't you delete this?

File: applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py [hunk 11]
```diff
@@ -0,0 +1,760 @@
...
+            "model_part_name" : "Thermic",
+            "domain_size" : -1,
+            "echo_level": 0,
+            "buffer_size": 2,
+            "analysis_type": "non_linear",
+            "model_import_settings": {
+                "input_type": "mdpa",
+                "input_filename": "unknown_name"
+            },
+            "restart_settings" : {
```
Comment (philbucher) [inline applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py]: this can be removed, this is done differently now
will also remove it from structural

File: applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py [hunk 12]
```diff
@@ -0,0 +1,760 @@
...
+            
+            KratosMultiphysics.ReplaceElementsAndConditionsProcess(self.main_model_part,self._get_element_condition_replace_settings()).Execute()
+        
+            self._set_and_fill_buffer()
+        
+        # This will be removed once the Model is fully supported! => It wont e necessary anymore
+        if not self.model.HasModelPart(self.main_model_part.Name):
+            self.model.AddModelPart(self.main_model_part)
+
+        print(self.model)
```
Comment (philbucher) [inline applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py]: without echo_lvl?

File: applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py [hunk 13]
```diff
@@ -0,0 +1,760 @@
...
+                convection_diffusion_solution_strategy.SetInitializePerformedFlag(True)
+            except AttributeError:
+                pass
+        self.Check()
+        self.print_on_rank_zero("::[ConvectionDiffusionBaseSolver]:: ", "Finished initialization.")
+
+    def GetOutputVariables(self):
+        pass
+
+    def SaveRestart(self):
```
Comment (philbucher) [inline applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py]: can be removed, is handled through a process now

File: applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py [hunk 14]
```diff
@@ -0,0 +1,760 @@
...
+            # SetInitializePerformedFlag is not a member of SolvingStrategy but
+            # is used by ResidualBasedNewtonRaphsonStrategy.
+            try:
+                convection_diffusion_solution_strategy.SetInitializePerformedFlag(True)
+            except AttributeError:
+                pass
+        self.Check()
+        self.print_on_rank_zero("::[ConvectionDiffusionBaseSolver]:: ", "Finished initialization.")
+
+    def GetOutputVariables(self):
```
Comment (philbucher) [inline applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py:302]: I would remove this

File: applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py [hunk 15]
```diff
@@ -0,0 +1,760 @@
...
+        if not hasattr(self, '_builder_and_solver'):
+            self._builder_and_solver = self._create_builder_and_solver()
+        return self._builder_and_solver
+
+    def get_convection_diffusion_solution_strategy(self):
+        if not hasattr(self, '_convection_diffusion_solution_strategy'):
+            self._convection_diffusion_solution_strategy = self._create_convection_diffusion_solution_strategy()
+        return self._convection_diffusion_solution_strategy
+
+    def get_restart_utility(self):
```
Comment (philbucher) [inline applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py]: can be removed

File: applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py [hunk 16]
```diff
@@ -0,0 +1,760 @@
...
+
+    def import_materials(self):
+        materials_filename = self.settings["material_import_settings"]["materials_filename"].GetString()
+        if (materials_filename != ""):
+            import read_materials_process
+            # Create a dictionary of model parts.
+            Model = KratosMultiphysics.Model()
+            Model.AddModelPart(self.main_model_part)
+            # Add constitutive laws and material properties from json file to model parts.
+            read_materials_process.ReadMaterialsProcess(Model, self.settings["material_import_settings"])
```
Comment (philbucher) [inline applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py]: cannot you pass `self.model`?

File: applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py [hunk 17]
```diff
@@ -0,0 +1,760 @@
...
+        step =-buffer_size
+        time = time - delta_time * buffer_size
+        self.main_model_part.ProcessInfo.SetValue(KratosMultiphysics.TIME, time)
+        for i in range(0, buffer_size):
+            step = step + 1
+            time = time + delta_time
+            self.main_model_part.ProcessInfo.SetValue(KratosMultiphysics.STEP, step)
+            self.main_model_part.CloneTimeStep(time)
+
+    def _get_restart_settings(self):
```
Comment (philbucher) [inline applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py]: can be removed

File: applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py [hunk 18]
```diff
@@ -0,0 +1,760 @@
...
+                            convection_diffusion_scheme,
+                            linear_solver,
+                            convection_diffusion_convergence_criterion,
+                            builder_and_solver,
+                            self.settings["max_iteration"].GetInt(),
+                            self.settings["compute_reactions"].GetBool(),
+                            self.settings["reform_dofs_at_each_step"].GetBool(),
+                            self.settings["move_mesh_flag"].GetBool())
+
+    def _create_restart_utility(self):
```
Comment (philbucher) [inline applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py]: can be removed

File: applications/convection_diffusion_application/tests/cpp_tests/test_example.cpp [hunk 19]
```diff
@@ -0,0 +1,39 @@
+// KRATOS ___ ___  _  ___   __   ___ ___ ___ ___
```
Comment (philbucher) [inline applications/convection_diffusion_application/tests/cpp_tests/test_example.cpp:1]: why did you add this file?

File: applications/convection_diffusion_application/tests/run_cpp_unit_tests.py [hunk 20]
```diff
@@ -0,0 +1,9 @@
+from KratosMultiphysics import *
```
Comment (philbucher) [inline applications/convection_diffusion_application/tests/run_cpp_unit_tests.py:1]: is this file needed?

**Review by philbucher — COMMENTED** [review]
some leftovers should be cleaned, then it is ok form my side

nice work :+1:

File: applications/convection_diffusion_application/custom_conditions/flux_condition.h (hunk 10 above)
Comment (philbucher) [reply applications/convection_diffusion_application/custom_conditions/flux_condition.h]: this is also referenced in many other files

Comment (jcotela) [reply applications/convection_diffusion_application/custom_conditions/flux_condition.h]: :+1: Please reference the standard `kratos/license.txt`

File: applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py (hunk 11 above)
Comment (RiccardoRossi) [reply applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py]: what do you mean by "differently"?

Comment (RiccardoRossi) [reply applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py]: oh you refer to the restart...

File: applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py (hunk 16 above)
Comment (RiccardoRossi) [reply applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py]: yes, actually we should NEVER (or almost so) create a new model...

File: applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py (hunk 12 above)
Comment (loumalouomega) [reply applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py]: Maybe there is a leftover from the debug

File: applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py (hunk 11 above)
Comment (loumalouomega) [reply applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py]: Well, I will check

File: applications/convection_diffusion_application/custom_conditions/flux_condition.h (hunk 10 above)
Comment (loumalouomega) [reply applications/convection_diffusion_application/custom_conditions/flux_condition.h]: Ok

File: applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py (hunk 16 above)
Comment (loumalouomega) [reply applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py]: I will check, maybe is a leftover from the first implementation

File: applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py (hunk 17 above)
Comment (loumalouomega) [reply applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py]: Yep

File: applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py (hunk 18 above)
Comment (loumalouomega) [reply applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py]: Ok

File: applications/convection_diffusion_application/tests/cpp_tests/test_example.cpp (hunk 19 above)
Comment (loumalouomega) [reply applications/convection_diffusion_application/tests/cpp_tests/test_example.cpp:1]: To check that the cpp test compiled on the app, can be removed

File: applications/convection_diffusion_application/tests/run_cpp_unit_tests.py (hunk 20 above)
Comment (loumalouomega) [reply applications/convection_diffusion_application/tests/run_cpp_unit_tests.py:1]: Well, eventually if some cpp test is added yes

**Comment by loumalouomega** [comment]
Done

File: applications/convection_diffusion_application/convection_diffusion_application.h [hunk 21]
```diff
@@ -77,10 +62,37 @@ KRATOS_DEFINE_3D_APPLICATION_VARIABLE_WITH_COMPONENTS( CONVECTION_DIFFUSION_APPL
 ///@name Kratos Classes
 ///@{
 
-/// Short class definition.
-/** Detail class definition.
+/** 
+ * @class KratosConvectionDiffusionApplication
+ * @ingroup KratosConvectionDiffusionApplication
+ * @brief The Convection DIffusion Application contains a series of elements and conditions and the corresponding strategies and solvers within Kratos Multiphysics necesaries in order to simulate a convection-diffusion problem
```
Comment (jcotela) [inline applications/convection_diffusion_application/convection_diffusion_application.h]: Minor typos: `DIffusion` `necesaries`

File: applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py [hunk 22]
```diff
@@ -0,0 +1,716 @@
...
+    """
+    def __init__(self, model, custom_settings):
+        super(ConvectionDiffusionBaseSolver, self).__init__(model, custom_settings)
+
+        default_settings = KratosMultiphysics.Parameters("""
+        {
+            "model_part_name" : "Thermic",
+            "domain_size" : -1,
+            "echo_level": 0,
+            "buffer_size": 2,
```
Comment (jcotela) [inline applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py]: Why is the buffer size a json option? I think this should not be settable by the user, since the choice of time integration scheme determines it.

File: applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py [hunk 23]
```diff
@@ -0,0 +1,716 @@
...
+            
+            # We set the properties that are nodal
+            self._assign_nodally_properties()
+            
+            materials_imported = True
+        else:
+            materials_imported = False
+        return materials_imported
+
+    def validate_and_transfer_matching_settings(self, origin_settings, destination_settings):
```
Comment (jcotela) [inline applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py]: Not for this PR, but this should probably be a separate utility (or maybe a feature of the base python solver class, @philbucher?)

File: applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py [hunk 24]
```diff
@@ -0,0 +1,716 @@
...
+    def __init__(self, model, custom_settings):
+        super(ConvectionDiffusionBaseSolver, self).__init__(model, custom_settings)
+
+        default_settings = KratosMultiphysics.Parameters("""
+        {
+            "model_part_name" : "Thermic",
+            "domain_size" : -1,
+            "echo_level": 0,
+            "buffer_size": 2,
+            "analysis_type": "non_linear",
```
Comment (jcotela) [inline applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py]: I'd default this to linear (I don't think that the ConvDiff element does anything non-linear by default)

File: applications/convection_diffusion_application/python_scripts/convergence_criteria_factory.py [hunk 25]
```diff
@@ -0,0 +1,55 @@
...
+import KratosMultiphysics
+
+# Check that applications were imported in the main script
+KratosMultiphysics.CheckRegisteredApplications("ConvectionDiffusionApplication")
+
+# Import applications
+import KratosMultiphysics.ConvectionDiffusionApplication as ConvectionDiffusionApplication
+
+# Convergence criteria class
+class convergence_criterion:
```
Comment (jcotela) [inline applications/convection_diffusion_application/python_scripts/convergence_criteria_factory.py]: Why is this in the convection-diffusion application? I see no application-specific things here...

File: applications/convection_diffusion_application/tests/cpp_tests/test_example.cpp [hunk 26]
```diff
@@ -0,0 +1,39 @@
...
+    namespace Testing 
+    {
+        /// Tests
+       
+        /** Example test
+         */
+        
+        KRATOS_TEST_CASE_IN_SUITE(ExampleTest, KratosConvectionDiffusionFastSuite) 
+        {
+            constexpr double tolerance = 1e-6;
```
Comment (jcotela) [inline applications/convection_diffusion_application/tests/cpp_tests/test_example.cpp]: This gives an unused variable warning.

File: applications/convection_diffusion_application/tests/run_cpp_unit_tests.py (hunk 20 above)
Comment (jcotela) [reply applications/convection_diffusion_application/tests/run_cpp_unit_tests.py:1]: I tend to agree with @philbucher, I'm not sure we need to add a cpp test runner in every application.

Maybe we could add the two lines of the run() function to the main test_ConvectionDiffusionApplication file?

File: applications/convection_diffusion_application/tests/test_ConvectionDiffusionApplication.py [hunk 27]
```diff
@@ -1,14 +1,34 @@
...
-from KratosMultiphysics import *
-from KratosMultiphysics.ConvectionDiffusionApplication import *
+import KratosMultiphysics
+import KratosMultiphysics.ConvectionDiffusionApplication
+import run_cpp_unit_tests
 
 # Import Kratos "wrapper" for unittests
 import KratosMultiphysics.KratosUnittest as KratosUnittest
 
+import subprocess
```
Comment (jcotela) [inline applications/convection_diffusion_application/tests/test_ConvectionDiffusionApplication.py]: Maybe a stupid question, but where are subprocesses used?

**Review by jcotela — CHANGES_REQUESTED** [review]
Thanks for the cleaning and the upgrade @loumalouomega. I'd be grateful if you could comment on the issues I have highlighted (since I may be wrong on some) before I approve.

File: applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py (hunk 22 above)
Comment (loumalouomega) [reply applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py]: We should add a method for the schemes to return the required buffer size. Besides, this is just what we do in the structural

File: applications/convection_diffusion_application/convection_diffusion_application.h (hunk 21 above)
Comment (loumalouomega) [reply applications/convection_diffusion_application/convection_diffusion_application.h]: OK

File: applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py (hunk 23 above)
Comment (loumalouomega) [reply applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py]: I agree to move to the base python solver

File: applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py (hunk 24 above)
Comment (loumalouomega) [reply applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py]: Ok

File: applications/convection_diffusion_application/tests/test_ConvectionDiffusionApplication.py (hunk 27 above)
Comment (loumalouomega) [reply applications/convection_diffusion_application/tests/test_ConvectionDiffusionApplication.py]: Huuuum, copy/paste from structural

File: applications/convection_diffusion_application/tests/cpp_tests/test_example.cpp (hunk 26 above)
Comment (loumalouomega) [reply applications/convection_diffusion_application/tests/cpp_tests/test_example.cpp]: True

**Comment by loumalouomega** [comment]
Done @jcotela

**Comment by loumalouomega** [comment]
Does something miss?

File: applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py (hunk 22 above)
Comment (jcotela) [reply applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py]: You are choosing the scheme already in the solvers, I would set the buffer size within the solver, in relation to the scheme choice (that is what we do in the fluid, at least)

File: applications/convection_diffusion_application/tests/test_ConvectionDiffusionApplication.py (hunk 27 above)
Comment (jcotela) [reply applications/convection_diffusion_application/tests/test_ConvectionDiffusionApplication.py]: So, can we remove it?

File: applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py (hunk 22 above)
Comment (loumalouomega) [reply applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py]: In structures we set manually, I was suggesting to add to scheme.h a method GetRequiredBuffer() or something like it, in order to not add manually each time the buffer required (needs to be defined on the scheme, or error can be triggered). Certainly I think this PR is already huge, so this can be done in a further one

File: applications/convection_diffusion_application/tests/test_ConvectionDiffusionApplication.py (hunk 27 above)
Comment (loumalouomega) [reply applications/convection_diffusion_application/tests/test_ConvectionDiffusionApplication.py]: I assume, @philbucher what do you think?

File: applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py (hunk 22 above)
Comment (jcotela) [reply applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py]: I would not add it to scheme.h, since it does not always know the buffer size (symbolic elements that integrate in time "internally" are an example of this). The python solver itself has all the information needed to determine the buffer size (scheme+element name), so it should not give the user a chance to set it to the wrong value...

File: applications/convection_diffusion_application/tests/test_ConvectionDiffusionApplication.py (hunk 27 above)
Comment (philbucher) [reply applications/convection_diffusion_application/tests/test_ConvectionDiffusionApplication.py]: yes pls remove

only needed for launching the mpi tests in the structural (which is for sure only a temp solution!)

Comment (loumalouomega) [reply applications/convection_diffusion_application/tests/test_ConvectionDiffusionApplication.py]: Ok

File: applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py (hunk 22 above)
Comment (loumalouomega) [reply applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py]: Any hint which size should I set?, I don't have that much experience with the Conv-Diff application

Comment (jcotela) [reply applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py]: I'd say 2 for dynamic problems (since this is enough both for Backward Euler and Crank-Nicolson). Probably 1 should be enough for static ones, but I admit I have never tried it.

Comment (loumalouomega) [reply applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py]: Well can be check easily, I added two tests, one static and other dynamic

**Comment by loumalouomega** [comment]
Done @jcotela

**Review by jcotela — APPROVED** [review]
